core/src/main/java/org/apache/iceberg/data/GenericDataUtil.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import java.nio.ByteBuffer;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.iceberg.util.DateTimeUtil;
+
+/** Utility methods for working with Iceberg's generic data model */
+public class GenericDataUtil {
+  private GenericDataUtil() {}
+
+  /**
+   * Convert a value from Iceberg's internal data model to the generic data model.
+   *
+   * @param type a data type
+   * @param value value to convert
+   * @return the value in the generic data model representation
+   */
+  public static Object internalToGeneric(Type type, Object value) {
+    if (null == value) {
+      return null;
+    }
+
+    switch (type.typeId()) {
+      case DATE:
+        return DateTimeUtil.dateFromDays((Integer) value);
```

Comment:
Value is an `Object`, so I cast to the object rather than the primitive. Looks like you're right that we can cast directly and unbox at the same time based on the method signature... but I typically don't do that for style reasons. If you cast to the primitive, then there's a risk of introducing an unnecessary failure if the value is `null` and the receiving function accepts an `Integer` that can be null. I think casting to the object version is better to avoid those problems.

Reply:
My reasoning is similar, but the other way around:

As you mentioned, the method accepts an `int`:
https://github.com/apache/iceberg/blob/91a1505d09cebcd1d088ac53cd42732c343883de/api/src/main/java/org/ap
ache/iceberg/util/DateTimeUtil.java#L49-L51

Casting to an `Integer` gives the impression to me that a `null` is acceptable here, while it is not. I'm not very strong on this one, let's pull in @nastra as a tiebreaker :)

Reply:
I agree with Ryan's reasoning. Casting to `int` will throw a NPE because it will first cast it to `Integer` and then throw the NPE when doing `Integer.intValue()`. That being said, it's safer to always cast to the object type instead of the primitive in case the value can be null